Some performance enhancements to avro and json serdes

rodesai opened the pull request:
### Description 
This patch implements some performance enhancements to the avro and json serde,
    listed in order of significance:

    - Don't deserialize records before passing them to the JsonConverter.
      Deserialization to json is very expensive. The JsonConverter already does this.
      The ksql json serde was deserializing to json to build the case insensitive name
      map. We can just do this using the value returned by JsonConverter.

    - Don't use the stream API from serdes. Turns out its actually pretty expensive.

    - Don't use struct.put/get(String fieldName). Struct has to then look up the Field
      by name. We already have the Field object handy, so we should use
      struct.put/get(Field field) instead.

    - Use ArrayList to build GenericRows.

    - Don't use varargs. They create a new array every time.

### Testing done 
No new functionality. Existing tests should catch regressions. Performance improvement verified experimentally.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

hjafarpour commented on ksql-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonDeserializer.java:
```diff
@@ -131,29 +118,32 @@ private Object enforceFieldType(final Schema fieldSchema, final Object columnVal
-        getCaseInsensitiveFieldNameMap(fieldSchema.fields());
     final Map<String, String> caseInsensitiveStructFieldNameMap =
-        getCaseInsensitiveStructFieldNameMap(structMap);
+        getCaseInsensitiveFieldNameMap(structMap, false);
```
If `omitAt` is always going to be `false`, do we need it?

rodesai replied:
Its true on line 82

hjafarpour commented on ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java:
```diff
@@ -71,14 +72,44 @@ private RuntimeException createTypeMismatchException(final String pathStr,
-        .orElseThrow(
-            () -> createTypeMismatchException(pathStr, schema, connectSchema));
+                            final Schema.Type[] validTypes) {
+    // don't use stream here
```
Long live `for` loop! ;)

miguno commented:
@rodesai:
> - Don't use the stream API from serdes. Turns out its actually pretty expensive.

Can you clarify?  Do you mean the Serdes of Kafka Streams?  If so, is there something we can do in Kafka Streams to improve this?  I am asking because serde work is often one of the most CPU intensive parts of any KStreams app.

rodesai commented:
@miguno stream API here refers to the java.util stream API: https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html

ijuma commented:
Do you have some numbers for these performance enhancements? Ideally, concrete numbers for each improvement as well as specifics of the benchmark would be provided.

rodesai commented:
The test measures throughput for a stateless KSQL query (CREATE STREAM SINK AS SELECT * FROM SRC) on an r4.xlarge instance with JSON and Avro data and the following schemas:

basic:
```
{
        "namespace": "ksql",
        "name": "perf_schema",
        "type": "record",
        "fields": [
                {"name": "long_field_1", "type": {
                    "type": "long",
                }},
                {"name": "long_field_2", "type": {
                    "type": "long",
                }},
                {"name": "string_field", "type": {
                    "type": "string",
                }}
        ]
}
```

twitter (avro only): https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/test/resources/schema-validation-tests/real-schema.json

After this patch we see a 7% improvement for JSON, an 18% improvement for basic avro, and a 28% improvement for avro with the twitter schema.

rodesai commented:
@apurvam - I don't have a microbenchmark yet. Still need to think through the right approach (try and run a more end-to-end test similar to the query validation tests, or just test the bits we know are hot spots (e.g. serdes)). I think it makes more sense to open a follow-up for that.

big-andy-coates commented on ksql-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonDeserializer.java:
```diff
@@ -131,29 +118,32 @@ private Object enforceFieldType(final Schema fieldSchema, final Object columnVal
-            e -> enforceFieldType(fieldSchema.valueSchema(), e.getValue())
-        ));
+      final Map<String, ?> columnMap) {
+    final Map<String, Object> ksqlMap = new HashMap<>();
```
consider setting initial capacity of the map.

big-andy-coates commented on ksql-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonDeserializer.java:
```diff
@@ -131,29 +118,32 @@ private Object enforceFieldType(final Schema fieldSchema, final Object columnVal
-        ));
+      final Map<String, ?> columnMap) {
+    final Map<String, Object> ksqlMap = new HashMap<>();
+    for (final Map.Entry<String, ?> e : columnMap.entrySet()) {
```
consider `columnMap.entrySet().forEach((k,v) -> ...`

big-andy-coates commented on ksql-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonDeserializer.java:
```diff
@@ -164,49 +154,23 @@ private Struct enforceFieldTypeForStruct(
-          keyMap.put(fieldName.toUpperCase(), fieldName);
+  private Map<String, String> getCaseInsensitiveFieldNameMap(final Map<String, ?> map,
+                                                             final boolean omitAt) {
+    final Map<String, String> keyMap = new HashMap<>();
```
Consider setting the map's initial capacity.

big-andy-coates commented on ksql-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonDeserializer.java:
```diff
@@ -164,49 +154,23 @@ private Struct enforceFieldTypeForStruct(
+  private Map<String, String> getCaseInsensitiveFieldNameMap(final Map<String, ?> map,
+                                                             final boolean omitAt) {
+    final Map<String, String> keyMap = new HashMap<>();
+    for (final Map.Entry<String, ?> entry : map.entrySet()) {
```
consider `columnMap.entrySet().forEach((k,v) -> ...`

rodesai commented:
I threw together a quick comparison of streams vs loops for small lists (which should be pretty common for ksql (e.g. each level of the schema has a handful of fields):

The output shows the result of a nonsense computation that iterates over a list, followed by the time taken in milliseconds. The row in each pair is for streams, the second row is for a loop. The code that uses streams takes substantially longer.

```
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

public class Main {

    public static void main(String[] args) {
        final int sz = 5;
        final long iters = 10000000;
        final Random r = new Random();

        for (int j = 0; j < 10; j++) {
            List<Integer> values = new LinkedList<>();
            for (int i = 0; i < sz; i++) {
                values.add(r.nextInt());
            }
            List<Integer> valuess = Collections.unmodifiableList(values);

            final long start = System.currentTimeMillis();
            long ignore = 0;
            final long jj = j;
            for (long i = 0; i < iters; i++) {
                ignore += valuess.stream()
                    .map(x -> x + jj)
                    .filter(x -> x % 2 == 0)
                    .collect(Collectors.toList()).size();
            }
            System.out.println(ignore + " " + (System.currentTimeMillis() - start));

            ignore = 0;
            final long start2 = System.currentTimeMillis();
            for (long i = 0; i < iters; i++) {
                final List<Integer> l = new ArrayList<>(valuess.size());
                for (Integer v : valuess) {
                    if ((v + j) % 2 == 0)
                        l.add(v + j);
                }
                ignore += Collections.unmodifiableList(l).size();
            }
            System.out.println(ignore + " " + (System.currentTimeMillis() - start2));

            System.out.println();
        }
    }
}
```

And get the following output:
```
40000000 2363
40000000 835

40000000 1663
40000000 774

20000000 1474
20000000 687

10000000 1512
10000000 603

20000000 1682
20000000 675

40000000 1665
40000000 740

20000000 1504
20000000 644

20000000 1421
20000000 659

10000000 1582
10000000 573

30000000 1530
30000000 705
```

ijuma commented:
You have to use jmh for micro benchmarks, otherwise your results can be very misleading. Particularly for abstractions that benefit from JIT compilation.

ijuma commented:
A couple more things:
1. When using primitives, boxing can play a major role in bad results, so that requires some care too.
2. Why are LinkedList and ArrayList mixed here? In Java, you almost always want the latter.

rodesai commented:
@ijuma was pretty sure I was missing some things - thanks for the pointers.

> Why are LinkedList and ArrayList mixed here? In Java, you almost always want the latter.

That was a bug.

rodesai commented:
Here's a JMH test that I think addresses the concerns with boxing/unboxing as well:

```
public class MyBenchmark {

    @State(Scope.Thread)
    public static class MyState {
        public List<Integer> values;
        public MyState() {
            List<Integer> l = new ArrayList<>(5);
            Random r = new Random();
            for (int i = 0; i < 5; i++) {
                l.add(r.nextInt() % 64);
            }
            values = Collections.unmodifiableList(l);
        }
    }

    @Benchmark
    public void testMethod(MyState s, Blackhole blackhole) {
        List<Integer> values = s.values;
        blackhole.consume(
            values.stream()
                .filter(x -> x % 2 == 0)
                .collect(Collectors.toList()));
    }

    @Benchmark
    public void testMethodLoop(MyState s, Blackhole blackhole) {
        List<Integer> values = s.values;
        final List<Integer> l = new ArrayList<>(values.size());
        for (Integer v : values) {
                if (v  % 2 == 0)
                        l.add(v);
        }
        blackhole.consume(l);
    }
}
```

Results:
```
Benchmark                    Mode  Cnt         Score         Error  Units
MyBenchmark.testMethod      thrpt   25   5896254.650 ±  973758.459  ops/s
MyBenchmark.testMethodLoop  thrpt   25  11151277.517 ± 1493611.419  ops/s
```

ijuma commented:
That's a good benchmark, thanks. I assume the numbers look closer if the size of the collection is 100? It seems reasonable that the additional allocations for Streams would make a measurable diffe

rodesai commented:
> I assume the numbers look closer if the size of the collection is 100? 

Yep - the overhead decreases as the collection grows.
